pkg/logql/functions.go

```diff
@@ -0,0 +1,63 @@
+package logql
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/prometheus/prometheus/promql"
+)
+
+const unsupportedErr = "unsupported range vector aggregation operation: %s"
+
+func (r rangeAggregationExpr) extractor() (SampleExtractor, error) {
+	switch r.operation {
+	case OpRangeTypeRate, OpRangeTypeCount:
+		return extractCount, nil
+	case OpRangeTypeBytes, OpRangeTypeBytesRate:
+		return extractBytes, nil
+	default:
+		return nil, fmt.Errorf(unsupportedErr, r.operation)
+	}
+}
+
+func (r rangeAggregationExpr) aggregator() (RangeVectorAggregator, error) {
+	switch r.operation {
+	case OpRangeTypeRate:
+		return rateLogs(r.left.interval), nil
+	case OpRangeTypeCount:
+		return countOverTime, nil
+	case OpRangeTypeBytesRate:
+		return rateLogBytes(r.left.interval), nil
+	case OpRangeTypeBytes:
+		return sumOverTime, nil
+	default:
+		return nil, fmt.Errorf(unsupportedErr, r.operation)
+	}
+}
+
+// rateLogs calculates the per-second rate of log lines.
+func rateLogs(selRange time.Duration) func(ts int64, samples []promql.Point) float64 {
```

Comment:
I notice none of the operations here actually use the `ts int64` argument. I think we can make this
`type RangeVectorAggregator func([]promql.Point) float64` WDYT?

Reply:
I did it.